File: AngularJS/src/org/angular2/web/Angular2SymbolOrigin.java
// Copyright 2000-2021 JetBrains s.r.o. and contributors. Use of this source code is governed by the Apache 2.0 license that can be found in the LICENSE file.
package org.angular2.web;

import com.intellij.javascript.nodejs.PackageJsonData;
import com.intellij.webSymbols.PsiSourcedWebSymbol;
import com.intellij.webSymbols.WebSymbolsContainer;
import com.intellij.lang.javascript.buildTools.npm.PackageJsonUtil;
import com.intellij.openapi.util.Pair;
import com.intellij.util.text.SemVer;
import icons.AngularJSIcons;
import org.angular2.Angular2Framework;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import javax.swing.*;

import static com.intellij.util.ObjectUtils.doIfNotNull;

public class Angular2SymbolOrigin implements WebSymbolsContainer.Origin {

  private final Angular2Symbol mySymbol;

  private Pair<String, String> nameAndVersion;

  public Angular2SymbolOrigin(Angular2Symbol symbol) { mySymbol = symbol; }

  private @NotNull Pair<String, String> getVersionAndName() {
    if (nameAndVersion == null) {
      var source = mySymbol instanceof PsiSourcedWebSymbol
                   ? ((PsiSourcedWebSymbol)mySymbol).getSource() : null;
      var psiFile = source != null ? source.getContainingFile() : null;
      var virtualFile = psiFile != null ? psiFile.getVirtualFile() : null;
      var pkgJson = virtualFile != null ? PackageJsonUtil.findUpPackageJson(virtualFile) : null;
      var data = pkgJson != null ? PackageJsonData.getOrCreate(pkgJson) : null;
      nameAndVersion = data != null
                       ? Pair.create(data.getName(), doIfNotNull(data.getVersion(), SemVer::toString))
                       : Pair.create(null, null);
    }
    return nameAndVersion;
  }

  @Nullable
  @Override
  public String getFramework() {
    return Angular2Framework.ID;
  }

  @Nullable
  @Override
  public String getPackageName() {
    return getVersionAndName().first;
  }

  @Nullable
  @Override
  public String getVersion() {
    return getVersionAndName().second;
  }

  @Nullable
  @Override
  public Icon getDefaultIcon() {
    return AngularJSIcons.Angular2;
  }
}

File: AngularJS/src/org/angular2/web/Angular2SymbolOrigin.kt
// Copyright 2000-2022 JetBrains s.r.o. and contributors. Use of this source code is governed by the Apache 2.0 license.
package org.angular2.web

import com.intellij.javascript.nodejs.PackageJsonData
import com.intellij.lang.javascript.buildTools.npm.PackageJsonUtil
import com.intellij.openapi.util.Pair
import com.intellij.webSymbols.PsiSourcedWebSymbol
import com.intellij.webSymbols.WebSymbolOrigin
import icons.AngularJSIcons
import org.angular2.Angular2Framework
import javax.swing.Icon

class Angular2SymbolOrigin(private val mySymbol: Angular2Symbol) : WebSymbolOrigin {

  private val versionAndName: Pair<String, String> by lazy(LazyThreadSafetyMode.PUBLICATION) {
    val source = if (mySymbol is PsiSourcedWebSymbol)
      (mySymbol as PsiSourcedWebSymbol).source
    else
      null
    val psiFile = source?.containingFile
    val virtualFile = psiFile?.virtualFile
    val pkgJson = if (virtualFile != null) PackageJsonUtil.findUpPackageJson(virtualFile) else null
    val data = if (pkgJson != null) PackageJsonData.getOrCreate(pkgJson) else null
    if (data != null)
      Pair.create(data.name, data.version?.toString())
    else
      Pair.create(null, null)
  }

  override val framework: String
    get() = Angular2Framework.ID

  override val library: String?
    get() = versionAndName.first

  override val version: String?
    get() = versionAndName.second

  override val defaultIcon: Icon
    get() = AngularJSIcons.Angular2
}
